CePO

pawelf-cerebras opened the pull request:
Adding Cerebras Planning and Optimization (CePO). On high-level, in CePO, we make m attempts to generate n step-by-step plans, refine the plans, check inconsistencies between them, use the above feedback to generate the final plan and produce the answer. This process is them repeated N times in a classical best of n manner.
- Added implementation of CePO
- Integrated CePO with optillm

pawelf-cerebras committed:
Update requirements to include Cerebras dependency

pawelf-cerebras committed:
Add support for CePO

pawelf-cerebras committed:
Initial version of CePO

pawelf-cerebras committed:
Add support for Cerebras client

pawelf-cerebras committed:
Add pairwise rating and clean up

pawelf-cerebras committed:
Fix default rating type

erich-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,297 @@
+@dataclass
+class CepoConfig:
+    bestofn_n: int = 3
+    bestofn_temperature: int = 0.1
```
These should not be `int`s but `float`s right?

erich-cerebras replied:
Fixed in my new PR

erich-cerebras commented:
Do we also want to maybe edit the `ReadMe.md` here: https://github.com/CerebrasResearch/cb_optillm/blob/cepo/README.md#implemented-techniques ?

erich-cerebras committed:
Add modification of CePO configs through yaml and cli arguments

This will allow users to do the following:
1. Cli where if they can pass in anything that's "cepo_<name-of-attribute>"
2. Yaml file where if they pass it in as "<name-of-attribute>"
3. If none of them have a specific attribute, we use the default setting
4. If both of them have the specific attribute, we error out

erich-cerebras committed:
Add modification of CePO configs through yaml and cli arguments

This will allow users to do the following:
1. Cli where if they can pass in anything that's "cepo_<name-of-attribute>"
2. Yaml file where if they pass it in as "<name-of-attribute>"
3. If none of them have a specific attribute, we use the default setting
4. If both of them have the specific attribute, we error out

erich-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,297 @@
+
+        try:
+            rating = float(rating_response)
+            ratings.append(rating)
```
Nit: Merge the two lines together since we don't need rating afterwards:
```
ratings.append(float(rating_response))
```

erich-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,297 @@
+        except ValueError:
+            ratings.append(0)
+        
+        rating_messages = rating_messages[:-2]
```
Can you add a comment on why you're using the magic number -2?

erich-cerebras reviewed (approved):
Just a couple of small comments, besides that it LGTM!

erich-cerebras committed:
Fix default CePO config yaml

erich-cerebras committed:
Merge pull request #2 from CerebrasResearch/erich/yaml_config

Add modification of CePO configs through yaml and cli arguments

pawelf-cerebras committed:
Fix check if cepo_config_file provided

codelion committed:
Convert multi-turn conversations back to messages

- remove the User: and Assistant: prefixes

codelion committed:
Update publish.yml

- Add official docker image to be published with release

codelion committed:
Update setup.py

Bump version for release

codelion committed:
Update setup.py

codelion committed:
Update setup.py

codelion committed:
Update setup.py

- bump version for new release

codelion committed:
Update readurls_plugin.py

codelion committed:
Update publish.yml

- Multi platform build

codelion committed:
Update setup.py

codelion committed:
Update readurls_plugin.py

pawelf-cerebras committed:
Add cepo to readme

pawelf-cerebras committed:
Minor cleaning

pawelf-cerebras committed:
Add VS Code to ignore list

ivanl-cerebras commented on configs/cepo_config.yaml:
```diff
@@ -0,0 +1,14 @@
+planning_max_tokens_step1: 4096
+planning_max_tokens_step2: 4096
+planning_max_tokens_step3: 4096
+planning_max_tokens_step4: 4096
```
Nit: this could be cleaner `planning_temperature_per_step` and `planning_max_tokens_per_step` lists

pawelf-cerebras replied:
Actually, this used to be a list, but we found it more convenient to have them as separate variables because we most often change only one of them for specific benchmarks. I'll keep it as is for now, but we may revert it back to list in the future. I agree it's a bit cleaner. Thanks for the suggestion.

ivanl-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,334 @@
+    planning_max_tokens_step1: int = 4096
+    planning_max_tokens_step2: int = 4096
+    planning_max_tokens_step3: int = 4096
+    planning_max_tokens_step4: int = 4096
```
It might be confusing when one could change the values here or in `configs/cepo_config.yaml`. Can we indicate in README that `configs/cepo_config.yaml` is the default config and remove defaults from dataclass to have a single source of truth?

pawelf-cerebras replied:
Thanks, Ivan, this is a good idea. I removed the defaults from  dataclass and use cepo_config.yaml instead, cmd line args will overwrite the defaults from the config.

ivanl-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,334 @@
+        if match:
+            rating_response = match.group(1)
+        else:
+            rating_response = "0"
```
nit: `rating_response = match.group(1) if match else "0"` is cleaner

pawelf-cerebras replied:
Agree, thanks!

ivanl-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,334 @@
+    elif cepo_config.bestofn_rating_type == "pairwise":
+        best_completion, completion_tokens_rating, cb_log = rate_completions_pairwise(system_prompt, initial_query, client, model, completions, cepo_config, cb_log)
+    else:
+        raise ValueError("Invalid rating type in cepo_config")
```
nit: 

a cleaner way could be
```

    if cepo_config.bestofn_rating_type == "absolute":
        rate_completions_fn = rate_completions_absolute
    elif cepo_config.bestofn_rating_type == "pairwise":
        rate_completions_fn = rate_completions_pairwise
    else:
        raise ValueError("Invalid rating type in cepo_config")

best_completion, completion_tokens_rating, cb_log = rate_completions_fn(system_prompt, initial_query, client, model, completions, cepo_config, cb_log)
```

pawelf-cerebras replied:
Yup, like it. Done. Thanks!

ivanl-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,334 @@
+        except ValueError:
+            ratings.append(0)
+        
+        rating_messages = rating_messages[:-2]  # remove the last two messages
```
nit: can the comment explain why we need to remove the messages

pawelf-cerebras replied:
Yeah, it's from their original implementation of BON, I updated the comment, which hopefully makes it clearer.

codelion committed:
Update Dockerfile

codelion committed:
Fix streaming response

- yield response to allow clients that expect streaming

codelion committed:
return a list

codelion committed:
prep for new release

codelion committed:
Fix the docker file path

- add dockerignore
- parameterize port in dockerfile

codelion committed:
prep for new release

codelion committed:
Update README.md

pawelf-cerebras committed:
Minor readibility improvements

rohand-cerebras commented on README.md:
```diff
@@ -196,54 +196,70 @@ response = client.chat.completions.create(
+| `--cepo_bestofn_rating_type`        | Type of rating in best of n stage ("absolute" or "pairwise")    | `"absolute"`    |
+| `--cepo_planning_n`                 | Number of plans generated in planning stage                     | 3               |
+| `--cepo_planning_m`                 | Number of attempts to generate n plans in planning stage        | 6               |
+| `--cepo_planning_temperature_step1` | Temperature for generator in step 1 of planning stage           | 0.55            |
+| `--cepo_planning_temperature_step2` | Temperature for generator in step 2 of planning stage           | 0.25            |
+| `--cepo_planning_temperature_step3` | Temperature for generator in step 3 of planning stage           | 0.1             |
+| `--cepo_planning_temperature_step4` | Temperature for generator in step 4 of planning stage           | 0               |
+| `--cepo_planning_max_tokens_step1`  | Maximum number of tokens in step 1 of planning stage            | 4096            |
+| `--cepo_planning_max_tokens_step2`  | Maximum number of tokens in step 2 of planning stage            | 4096            |
+| `--cepo_planning_max_tokens_step3`  | Maximum number of tokens in step 3 of planning stage            | 4096            |
+| `--cepo_planning_max_tokens_step4`  | Maximum number of tokens in step 4 of planning stage            | 4096            |
```
Do we define what step 1-4 are? I don't see it in the readme.

pawelf-cerebras replied:
I added description of how CePO works to the README. Thanks for suggestion!

rohand-cerebras commented on optillm/cepo.py:
```diff
@@ -0,0 +1,333 @@
+
+@dataclass
+class CepoConfig:
+    bestofn_n: int = 3
+    bestofn_temperature: float = 0.1
+    bestofn_max_tokens: int = 4096
+    bestofn_rating_type: Literal["absolute", "pairwise"] = "absolute"
+    planning_n: int = 3
+    planning_m: int = 6
+    planning_temperature_step1: float = 0.55
+    planning_temperature_step2: float = 0.25
+    planning_temperature_step3: float = 0.1
+    planning_temperature_step4: float = 0
+    planning_max_tokens_step1: int = 4096
+    planning_max_tokens_step2: int = 4096
+    planning_max_tokens_step3: int = 4096
+    planning_max_tokens_step4: int = 4096
```
Could we document what each of these params do? ex: for those not familiar with n,m, and step 1-4.

pawelf-cerebras replied:
Yes, good point. I added description from README to the dataclass, so it's hopefully easier now. We'll also add description of the overall flow.

pawelf-cerebras committed:
Removed unnecessary comment

codelion committed:
add gui using new gradio app

codelion committed:
Update setup.py

prepare for new release

pawelf-cerebras committed:
Add cepo results

codelion committed:
Use modernbert based router

codelion committed:
Update setup.py

prepare for new release

pawelf-cerebras committed:
Make cepo_config.yaml define the default values instead of the dataclass for single source of truth

pawelf-cerebras committed:
Updated documentation of CePO

pawelf-cerebras committed:
Add description of CePO method

pawelf-cerebras committed:
Update CePO section of README

codelion committed:
Update optillm.py

codelion committed:
Update eval_aime_benchmark.py

- handle multiple responses as multiple attempts

codelion committed:
Update eval_aime_benchmark.py

pawelf-cerebras committed:
Add results for LiveCodeBench and SimpleQA

codelion committed:
Update setup.py

- prepare for new release

codelion committed:
Update README.md

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+import re
```
@pawelf-cerebras I think we should add LICENSE for the cepo code. In the past we've done something like this.

https://github.com/CerebrasResearch/Sparse-IFT/blob/master/cbsparse/LICENSE

vithursant replied:
@pawelf-cerebras did you have any thoughts on this?

pawelf-cerebras replied:
Yes, thanks for pointing this out. We're going to add the license. I mentioned this last week to Emma and she's making sure with the legal team that everything is good. I want to leave this comment unresolved until I hear from her and add the license.

vithursant replied:
Great, thanks!

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+    return completions[best_index], completion_tokens, cb_log
+
+
+def cepo(system_prompt: str, initial_query: str, client, model: str, cepo_config: Optional[CepoConfig] = None) -> list[str, int, dict]:
```
The cepo function is annotated to return list[str, int, dict], but it actually returns a tuple containing a str, int, dict. Update the return type annotation to match the actual return values.

pawelf-cerebras replied:
Good catch! Corrected

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+
+    # get the yaml file arguments
+    cepo_config_yaml = {}
+    if cmd_line_args["cepo_config_file"]:
```
The function accesses cmd_line_args["cepo_config_file"] directly, which can raise a KeyError if the key is absent.

Use .get() with a default value to safely access dictionary keys.

```
cepo_config_file = cmd_line_args.get("cepo_config_file")
if cepo_config_file:
    with open(cepo_config_file, "r") as yaml_file:
        cepo_config_yaml = yaml.safe_load(yaml_file)
```

pawelf-cerebras replied:
In the current implementation, cmd_line_args will always have "cepo_config_file", but as this is outside cepo module I agree this is good practice. Thanks for the suggestion! Done

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+import yaml
+
+@dataclass
+class CepoConfig:
```
The CepoConfig dataclass does not have default values, making it mandatory to provide all fields during instantiation. Additionally, the cepo function attempts to instantiate CepoConfig() without arguments, which will raise a TypeError. 

I propose we provide default values for optional configurations or make cepo_config a required parameter.

pawelf-cerebras replied:
Originally, we had default values, but I followed Ivan's suggestion to remove them and have the single source of truth in the config yaml file. I made cepo_config required in cepo function and removed the CepoConfig() call. Good catch!

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+    return question_only
+
+
+def generate_completion(system_prompt: str, task: str, client, model: str, cepo_config: CepoConfig) -> str:
```
The client parameter lacks a specific type hint, making it unclear what type of object is expected. If a specific type isn't available, consider using `Any` with appropriate documentation.

```
from typing import Any

def generate_completion(..., client: Any, ...):
    """`client` should be an instance of OpenAIClient or similar."""
    ...
```

pawelf-cerebras replied:
I followed the style they use in optillm for other methods (e.g. see bon.py). What is the benefit of using Any over omitting type?

vithursant replied:
okay, we can stick to the optillm style to maintain consistency. No real benefit besides maintaining the type of object for every argument.

pawelf-cerebras replied:
I actually updated the type as I agree it looks more consistent this way. I was just curious if there is any benefit that I'm missing. Thanks, Vithu!

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+    """We noticed that sometimes if the task includes specific formatting instructions, they may interfere with the reasoning flow. This
+    is a temporary workaround to extract the question only from the task. Work in progress.
+    """
+    question_only = task.replace('\n## Question: \n\n', '')
```
The function uses hard-coded string replacements that are brittle and may fail if the input format changes even slightly. Maybe regular expressions or more flexible parsing methods to extract the question. Alternatively, we can clarify and enforce the expected input format.

```
import re

def extract_question_only(task: str) -> str:
    """Extracts the question from the task string."""
    match = re.search(r'\n## Question:\s*(.*?)\n', task, re.DOTALL)
    if match:
        return match.group(1).strip()
    return task.strip()
```

pawelf-cerebras replied:
This is only temporary and specifically for CRUX and Math-L5 only. We will remove it entirely hopefully very soon, so I'll keep it for now as is, thanks for the suggestion though.

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+        )
+        final_solution = response.choices[0].message.content
+        completion_tokens += response.usage.completion_tokens
+    except (cerebras.cloud.sdk.BadRequestError, openai.BadRequestError) as e:
```
The function catches specific exceptions (`cerebras.cloud.sdk.BadRequestError` and `openai.BadRequestError`) but lacks imports for these exceptions, potentially leading to NameError.

pawelf-cerebras replied:
Another good catch! Thanks! I must have missed these when moving from our internal repo. Corrected

vithursant commented on optillm/cepo.py:
```diff
@@ -0,0 +1,314 @@
+    return question_only
+
+
+def generate_completion(system_prompt: str, task: str, client, model: str, cepo_config: CepoConfig) -> str:
```
Several functions lack comprehensive docstrings, making it harder for others to understand their purpose and usage.

E.g., 

```
def generate_completion(system_prompt: str, task: str, client: Any, model: str, cepo_config: CepoConfig) -> Tuple[str, int, dict]:
    """
    Generates a completion based on the provided system prompt and task.

    Parameters:
        system_prompt (str): The system prompt to guide the model.
        task (str): The task or question to be addressed.
        client (Any): The client instance for interacting with the AI model.
        model (str): The model name to be used for generating completions.
        cepo_config (CepoConfig): Configuration parameters for the CEPo process.

    Returns:
        Tuple[str, int, dict]: The generated completion, number of tokens used, and a log dictionary.
    """
    ...
```

pawelf-cerebras replied:
Good point, I added docstrings to all functions. Thanks!

vithursant reviewed (commented):
Left some comments, overall looking good.

codelion committed:
Update README.md

pawelf-cerebras committed:
Correct type of the output of cepo

vithursant reviewed (approved):
lgtm

pawelf-cerebras committed:
Minor fixes and add docstrings

pawelf-cerebras commented:
Thank you all for your time and feedback! That was very helpful! We're only missing the license and when it's in, we're good to merge.

emmac-cerebras committed:
Updatee README.md to add the discord/research channel link

pawelf-cerebras committed:
Merge pull request #3 from emmac-cerebras/patch-1

Updated README.md to add the discord/research channel link

emmac-cerebras committed:
Create NOTICE.md

emmac-cerebras committed:
Update .gitignore

Added a license header

emmac-cerebras committed:
Update NOTICE.md

Added a link to view the file changes

emmac-cerebras committed:
Update optillm.py

Added a license header

emmac-cerebras committed:
Update cepo.py

Added a license header

pawelf-cerebras committed:
Fix typo

pawelf-cerebras committed:
Make cepo_config required parameter

codelion committed:
Update eval_arena_hard_auto_rtc.py

- update max tokens

codelion committed:
Update coc_plugin.py

codelion committed:
Update coc_plugin.py

fix bug

codelion committed:
fix dependenices on GPU for local inference

add protobuf

codelion committed:
update dependencies

codelion committed:
Update inference.py

fix inference on amd gpu

codelion committed:
Revert "Update inference.py"

This reverts commit c0f0893c5b502c935181284f0f2b2e46395fae74.

codelion committed:
Update publish.yml

codelion committed:
add plugindir flag and env var

codelion committed:
Update publish.yml

codelion committed:
Update eval_aime_benchmark.py
